### PR TITLE
Added support for xfinity uei xhk1 ue

### PR DESCRIPTION
**Pull request by dettrick**

### CODE REVIEW

File: converters/toZigbee.js [hunk 1]
```diff
@@ -1242,6 +1242,26 @@ const converters = {
             entity.commandResponse('ssIasAce', 'panelStatusChanged', payload);
         },
     },
+    // set the status of keypad in accordance with the assigned Zigbee standard IAS status'
+    set_status: {
+        key: ['set_status'],
+        convertSet: async (entity, key, value, meta) => {
+            const panelStatus = utils.getKeyByValue(common.panStat, value.panelstatus, undefined);
+            const secondsRemain = value.secondsremain = 0;
```
Comment (Koenkk) [inline converters/toZigbee.js:1250]: I've never seen this syntax before, what does it do?

File: converters/toZigbee.js [hunk 2]
```diff
@@ -1242,6 +1242,26 @@ const converters = {
             entity.commandResponse('ssIasAce', 'panelStatusChanged', payload);
         },
     },
+    // set the status of keypad in accordance with the assigned Zigbee standard IAS status'
+    set_status: {
+        key: ['set_status'],
+        convertSet: async (entity, key, value, meta) => {
+            const panelStatus = utils.getKeyByValue(common.panStat, value.panelstatus, undefined);
```
Comment (Koenkk) [inline converters/toZigbee.js:1249]: Please make everything in value snake case, e.g. `panelStatus` -> `panel_status`.

File: devices.js [hunk 3]
```diff
@@ -8511,6 +8511,39 @@ const devices = [
...
+        zigbeeModel: ['URC4450BC0-X-R'],
+        model: 'URC4450BC0-X-R',
+        vendor: 'Xfinity',
+        description: 'Alarm Keypad',
+        supports: 'action, arm',
+        meta: {configureKey: 1, commandArmIncludeTransaction: true},
+        fromZigbee: [fz.command_arm, fz.temperature, fz.battery, fz.ias_occupancy_alarm_1, fz.identify, fz.ias_contact_alarm_1],
+        exposes: [e.battery(), e.voltage(), e.occupancy(), e.battery_low(), e.tamper(), e.presence(), exposes.numeric('action_code', 'r'),
+            exposes.numeric('action_zone', 'r'), e.temperature(), e.action([
+                'disarm', 'arm_day_zones', 'identify', 'arm_night_zones', 'arm_all_zones', 'invalid_code', 'emergency',
```
Comment (Koenkk) [inline devices.js:8524]: Please add `set_status` here, should be something like:


```js
new Composite('set_status', 'set_status')
  .withFeature(new exposes.Numeric('seconds_remaining', access.SET))
  .withFeature(new exposes.Numeric('alarm_status', access.SET))
  .withFeature(new exposes.Enum('panel_status', access.SET, ['arm', '...... more'])),
  . withDescription('Set status of the panel');
```

Comment (dettrick) [reply devices.js:8524]: I've tried to add but get the following error

image

Comment (Koenkk) [reply devices.js:8524]: Fixed in latest dev. (https://www.zigbee2mqtt.io/how_tos/how-to-switch-to-dev-branch.html)

Comment (dettrick) [reply devices.js:8524]: I've switched to dev branch and it now starts, however no controls are presented in home assistant and in the zigbee2mqtt web interface it shows the following for exposes

image

Comment (Koenkk) [reply devices.js:8524]: @nurikk I think composite is not supported in the frontend yet (since we did not use it before). Can this be added?

Comment (nurikk) [reply devices.js:8524]: Just merged into dev branch, welcome to test

Comment (dettrick) [reply devices.js:8524]: It now comes up in the exposes page and i can set the status from that page, still nothing in the actual home assistant entities section though
image

Comment (Koenkk) [reply devices.js:8524]: Home Assistant does not supported such advanced things so it will only be available through the frontend.

**Comment by cjsteward** [comment]
@Koenkk @dettrick Hey guys, is there anything else that needs to be done for this before it can be merged? I'm not a developer, but I've dabbled. If there's anything I can help with to get this merged please let me know. This integration is something I've been looking forward to for a long time now

**Comment by Multipage** [comment]
@Koenkk @dettrick When will this be merged? I'm trying to merge this with the dev version of zigbee2mqtt and can't get it to work as it should.
Also : why is there a common.js when there is no common.js in the branch. Am i missing something here?

File: converters/toZigbee.js (hunk 1 above)
Comment (dettrick) [reply converters/toZigbee.js:1250]: This one set's the indicator LED's and the beeping tones for the keypad. It's different to arm mode which is simply keeps the keypad mode in sync with the server mode but doesn't actually change any indications on the panel

File: devices.js (hunk 3 above)
Comment (dettrick) [reply devices.js:8524]: Ok will have to wait until it is.

**Comment by dettrick** [comment]
Ok not sure how to close this one out now. Can someone please help

**Comment by Multipage** [comment]
@dettrick  I'm sorry to bother you but i have some changes for your code but since i can't get it to work on dev right now because of troubles on my setup(i think) maybe you can take a look at it and get you on track faster?

In devices.js:
await bind(endpoint, coordinatorEndpoint, clusters);
await configureReporting.temperature(endpoint);
await configureReporting.batteryVoltage(endpoint);

Must be:
await reporting.bind(endpoint, coordinatorEndpoint, clusters);
await reporting.temperature(endpoint);
await reporting.batteryVoltage(endpoint);

The changes you made to common.js must now be done in constants.js

The set_status must be ..i think..from:
`
    set_status: {
        key: ['set_status'],
        convertSet: async (entity, key, value, meta) => {
            const panelStatus = utils.getKeyByValue(common.panStat, value.panelstatus, undefined);
            const secondsRemain = value.secondsremain = 0;
            const audibleNotif = value.audiblenotif = 0;
            const alarmStatus = value.alarmstatus = 0;

            if (panelStatus === undefined) {
                throw new Error(
                    `Unsupported status: '${value.panelstatus}', should be one of: ${Object.values(common.panStat)}`,
                );
            }

            globalStore.putValue(entity, 'panelStatus', panelStatus);
            const payload = {panelstatus: panelStatus, secondsremain: secondsRemain, audiblenotif: audibleNotif, alarmstatus: alarmStatus};
            entity.commandResponse('ssIasAce', 'panelStatusChanged', payload);
        },
    },            
                        
`
to:
`   set_status: {
        key: ['set_status'],
        convertSet: async (entity, key, value, meta) => {
            const panel_status = utils.getKey(constant.panStat, value.panelstatus, undefined);
            const seconds_remain = value.secondsremain = 0;
            const audible_notif = value.audiblenotif = 0;
            const alarm_status = value.alarmstatus = 0;

            if (panelStatus === undefined) {
                throw new Error(
                    `Unsupported status: '${value.panelstatus}', should be one of: ${Object.values(constant.panStat)}`,
                );
            }

            globalStore.putValue(entity, 'panelStatus', panel_status);
            const payload = {panelstatus: panel_status, secondsremain: seconds_remain, audiblenotif: audible_notif, alarmstatus: alarm_status};
            entity.commandResponse('ssIasAce', 'panelStatusChanged', payload);
        },
    },`

**Comment by dettrick** [comment]
Thanks i've added this to the new PR

**Comment by Multipage** [comment]
The PR isn't done yet so that's why i will post my findings here. I'll hope you don't mind and i hop it's usefull

Another fix:
const panel_status = utils.getKey(constant.panStat, value.panelstatus, undefined);
must be:
const panel_status = utils.getKey(constant.panStat, value.panelstatus, undefined, Number);


Currently on the last dev version i get this error when i try to set the status through MQTT:
`Zigbee2MQTT:error` 2021-01-20 13:20:01: Failed to call 'EntityPublish' 'onMQTTMessage' (TypeError: Cannot read property 'key' of undefined
    at /app/lib/extension/publish.js:185:56
    at Array.find (<anonymous>)
    at EntityPublish.onMQTTMessage (/app/lib/extension/publish.js:185:42)
    at Controller.callExtensionMethod (/app/lib/controller.js:382:44)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5))
 `
    
 I can't figure out what's wrong but maybe you can!

zigbee2mqtt version: current dev
herdsman-converters version  14.0.25

**Comment by GSzabados** [comment]
I have look all along what @dettrick has proposed, and I think that the arm_mode and panel_status should have been separated as he suggested. There is a decent overlap between the two, but they are not the same. 
The most obvious difference is the `invalid_code` vs `exit_delay`. If an `exit_delay` is sent to the panel with a `/zigbee2mqtt/FRIENDLYNAME/set` command when a wrong code has been entered the panel will react as it has received `ArmNotification` as `E_CLD_IASACE_ARM_NOTIF_INVALID_ARM_DISARM_CODE` according to the _ZigBee Cluster Library User Guide_, but as the arm_mode was set, then next time when the panel asks for panelStatus the `exit_delay` will be responded, and the panel will show _Exit Delay_.

I know there is a bit quirky solution to fix this outside of Z2M as after sending the 'exit_delay' the last arm mode has to be resent as well. But @dettrick's suggested way was providing a full solution for panel status and arm mode.

**Comment by thecobra666** [comment]
> I have look all along what @dettrick has proposed, and I think that the arm_mode and panel_status should have been separated as he suggested. There is a decent overlap between the two, but they are not the same.
> The most obvious difference is the `invalid_code` vs `exit_delay`. If an `exit_delay` is sent to the panel with a `/zigbee2mqtt/FRIENDLYNAME/set` command when a wrong code has been entered the panel will react as it has received `ArmNotification` as `E_CLD_IASACE_ARM_NOTIF_INVALID_ARM_DISARM_CODE` according to the _ZigBee Cluster Library User Guide_, but as the arm_mode was set, then next time when the panel asks for panelStatus the `exit_delay` will be responded, and the panel will show _Exit Delay_.
> 
> I know there is a bit quirky solution to fix this outside of Z2M as after sending the 'exit_delay' the last arm mode has to be resent as well. But @dettrick's suggested way was providing a full solution for panel status and arm mode.

Would love for this to be implemented. There is also a difference when the keypad is used with ZHA. In ZHA when pushing the one of the zone buttons, they light up. So when typing in the code, you know what zone is going to be activated. I don't know how they do it because in the  mqtt messages in Z2M I'm unable to get the zone status before the arming code has been typed in.

When I first go to the keypad the following message gets send:

`debug 2021-06-16 07:55:15: Received Zigbee message from '0x000d6f0014a6fe0a', type 'commandGetPanelStatus', cluster 'ssIasAce', data '{}' from endpoint 1 with groupID null
info  2021-06-16 07:55:15: MQTT publish: topic 'zigbee2mqtt/0x000d6f0014a6fe0a', payload '{"action":null,"action_code":null,"action_zone":null,"battery":100,"battery_low":false,"contact":true,"linkquality":255,"occupancy":true,"presence":null,"tamper":false,"temperature":24,"voltage":7000}'`

Then when I have selected the zone and entered the code, this gets send.

`debug 2021-06-16 07:55:17: Received Zigbee message from '0x000d6f0014a6fe0a', type 'commandArm', cluster 'ssIasAce', data '{"armmode":3,"code":"9999","zoneid":23}' from endpoint 1 with groupID null
info  2021-06-16 07:55:17: MQTT publish: topic 'zigbee2mqtt/0x000d6f0014a6fe0a', payload '{"action":"arm_all_zones","action_code":"9999","action_zone":23,"battery":100,"battery_low":false,"contact":true,"linkquality":255,"occupancy":true,"presence":null,"tamper":false,"temperature":24,"voltage":7000}'
info  2021-06-16 07:55:17: MQTT publish: topic 'zigbee2mqtt/0x000d6f0014a6fe0a', payload '{"action":"","action_code":null,"action_zone":null,"battery":100,"battery_low":false,"contact":true,"linkquality":255,"occupancy":true,"presence":null,"tamper":false,"temperature":24,"voltage":7000}'
info  2021-06-16 07:55:17: MQTT publish: topic 'zigbee2mqtt/0x000d6f0014a6fe0a/action', payload 'arm_all_zones'`

So I wonder how they intercept the zone option. Would anyone know? Alsi I'm unable to disable the entry or exit_delay mod e by arming it with a zone. The only way to stop the keypad from beeping is sending a disarm code.

**Pull request closed without merging**